### ISSUE TITLE
Extend flowistry to types with interior mutability

### ISSUE THREAD

**Issue by jyn514** [willcrichton/flowistry#53]
You mentioned two things during your thesis defense:
1. There are two analyses that flowistry supports: whole-program analysis, and (very accurate) heuristics using the lifetime annotations people already provide for the borrow checker.
2. The lifetime annotations only work for exterior mutability; things like `AtomicUsize::set` are not considered by flowistry to affect the data flow of the program.

I think it would be possible to extend this to interior mutability by using the intra-procederal analysis that looks into dependencies, but only for types which have interior mutability. The compiler already knows statically which types have interior mutability, because they have to contain an `UnsafeCell` (anything else is already undefined behavior).

**Comment by WaffleLapkin** [comment]
> The compiler already knows statically which types have interior mutability, because they have to contain an `UnsafeCell` (anything else is already undefined behavior).

That's true only for owned types though, something like `NonNull<T>` or `*mut T` can also mutate the value.

**Comment by jyn514** [comment]
@WaffleLapkin those types aren't using interior mutability, they're using unsafe. Writing to a `*mut T` when you only have a `&T` is undefined behavior.

**Comment by WaffleLapkin** [comment]
@jyn514 pointers can point to the heap, they not necessarily come from `&T` . Or they can actually point to an `UnsafeCell` somewhere, they were just `cast`ed to point to `T` (that's ok because `UnsafeCell` is `repr(transparent)`).
